src/Smalot/PdfParser/PDFObject.php

```diff
@@ -184,14 +184,49 @@ public function cleanContent(string $content, string $char = 'X')
         return $content;
     }
 
+    /**
+     * Add Q & q flags and Tf commands which before text block.
+     *
+     * @see: https://github.com/smalot/pdfparser/issues/387
+     * @see: https://github.com/smalot/pdfparser/issues/542
+     */
+    private function addQAndqFlagsAndTfCommands(string $section, $matches, int $pos): string
+    {
+        if (preg_match_all('/(?:\s|^)([Qq])(?:\s|$)/', $matches[1][$pos][0], $qMatches, \PREG_OFFSET_CAPTURE)) {
+            $len = \strlen($matches[1][$pos][0]);
+            $matchesCount = count($qMatches[0]);
+            for ($i = $matchesCount - 1; $i >= 0; --$i) {
+                $str = substr($matches[1][$pos][0], $qMatches[0][$i][1] + 3, $len - ($qMatches[0][$i][1] + 3));
+                $len = $qMatches[0][$i][1];
+                if (preg_match('/\sTf(\s|$)/', $str)) {
+                    $section = trim($str)."\n".$section;
+                }
+
+                if ('Q' == $qMatches[1][$i][0]) {
+                    $section = "Q\n".$section;
+                } elseif ('q' == $qMatches[1][$i][0]) {
+                    $section = "q\n".$section;
+                }
+            }
+            $str = substr($matches[1][$pos][0], 0, $qMatches[0][0][1]);
+            if (preg_match('/\sTf(\s|$)/', $str)) {
+                $section = trim($str)."\n".$section;
+            }
+        } elseif (preg_match('/\sTf(\s|$)/', $matches[1][$pos][0])) {
+            $section = trim($matches[1][$pos][0])."\n".$section;
+        }
+
+        return $section;
```

Comment:
I moved this to a separate function. 

If possible, add the comments you made here also in code. That would bolster code readability and understanding.